epochStart/bootstrap/epochStartDataProvider.go

```diff
@@ -0,0 +1,512 @@
+	esdp.requestEpochStartMetaBlock(epoch)
+
+	time.Sleep(delayAfterRequesting)
+
+	for {
+		numConnectedPeers := len(esdp.messenger.Peers())
+		threshold := int(thresholdForConsideringMetaBlockCorrect * float64(numConnectedPeers))
+		mb, errConsensusNotReached := esdp.epochStartMetaBlockInterceptor.GetEpochStartMetaBlock(threshold, epoch)
+		if errConsensusNotReached == nil {
+			return mb, nil
+		}
+		log.Info("consensus not reached for meta block. re-requesting and trying again...")
+		esdp.requestEpochStartMetaBlock(epoch)
+	}
+}
+
+func (esdp *epochStartDataProvider) getShardCoordinator(metaBlock *block.MetaBlock, nodesConfig *sharding.NodesSetup) (sharding.Coordinator, error) {
+	shardID, err := esdp.getShardID(nodesConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	numOfShards := len(metaBlock.EpochStart.LastFinalizedHeaders)
+	if numOfShards == 1 {
+		return &sharding.OneShardCoordinator{}, nil
+	}
+
+	return sharding.NewMultiShardCoordinator(uint32(numOfShards), shardID)
+}
+
+func (esdp *epochStartDataProvider) getShardHeaders(
+	metaBlock *block.MetaBlock,
+	nodesConfig *sharding.NodesSetup,
+	shardCoordinator sharding.Coordinator,
+) (map[uint32]*block.Header, error) {
+	headersMap := make(map[uint32]*block.Header)
+
+	shardID := shardCoordinator.SelfId()
+	if shardID == core.MetachainShardId {
+		for _, entry := range metaBlock.EpochStart.LastFinalizedHeaders {
+			var hdr *block.Header
+			hdr, err := esdp.getShardHeader(entry.HeaderHash, entry.ShardID)
+			if err != nil {
+				return nil, err
+			}
+			headersMap[entry.ShardID] = hdr
+		}
+
+		return headersMap, nil
+	}
+
+	var entryForShard *block.EpochStartShardData
+	for _, entry := range metaBlock.EpochStart.LastFinalizedHeaders {
+		if entry.ShardID == shardID {
+			entryForShard = &entry
+		}
+	}
+
+	if entryForShard == nil {
+		return nil, errors.New("shard data not found")
+	}
+
+	hdr, err := esdp.getShardHeader(
+		entryForShard.HeaderHash,
+		entryForShard.ShardID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	headersMap[shardID] = hdr
+	return headersMap, nil
+}
+
+func (esdp *epochStartDataProvider) getShardHeader(
+	hash []byte,
+	shardID uint32,
+) (*block.Header, error) {
+	esdp.requestShardHeader(shardID, hash)
+	time.Sleep(delayBetweenRequests)
+
+	for {
+		numConnectedPeers := len(esdp.messenger.Peers())
+		threshold := int(thresholdForConsideringMetaBlockCorrect * float64(numConnectedPeers))
+		mb, errConsensusNotReached := esdp.shardHeaderInterceptor.GetShardHeader(threshold)
+		if errConsensusNotReached == nil {
+			return mb, nil
+		}
+		log.Info("consensus not reached for shard header. re-requesting and trying again...")
+		esdp.requestShardHeader(shardID, hash)
+	}
+}
+
+func (esdp *epochStartDataProvider) requestMetaBlock(hash []byte) {
+	// send more requests
+	log.Debug("requested meta block", "hash", hash)
+	for i := 0; i < numRequestsToSendOnce; i++ {
+		time.Sleep(delayBetweenRequests)
+		esdp.requestHandlerMeta.RequestMetaHeader(hash)
```

Comment:
requestHandler would be enough as variable name.

Reply:
I wanted to suggest that the used requestHandler is for metachain

Reply:
but renamed